File: face/geomajas-face-dojo/client/src/main/resources/geomajas/map/print/PrintTemplate.js
dojo.provide("geomajas.map.print.PrintTemplate");
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */
dojo.declare("PrintTemplate", null, {

	/**
	 * @class 
	 * A print template (mirror of server object)
	 * @author Jan De Moerloose
	 *
	 * @constructor
	 */
	constructor : function(/* json object */ json) {
		log.info("   creating template: "+ (json ? json.name : "new"));
		if(json){
			dojo.mixin(this,json);
			this._createPage();
		}
	},
	
	configureMapModel : function(/* MapModel */ mapModel) {
		var mapId = mapModel.getId();
		var map = this.getMapComponent(mapId);
		var legend = this.getLegendComponent(mapId);
		map.copyLayers(mapModel);
		legend.copyLayers(mapModel);
	},
	
	/**
	 * find the map with the specified id in the descendancy tree
	 */
	getMapComponent : function(/* map id */ mapId) {
		var map = null;
		this.page.executeTopDown(
			function(component) {
				if(component.declaredClass == "MapComponent"){
					if(component.mapId == mapId || component.tag == "map"){
						map = component;
					}
				}
			}
		);
		return map;
	},

	/**
	 * find the map with the specified id in the descendancy tree
	 */
	getLegendComponent : function(/* map id */ mapId) {
		var legend = null;
		this.page.executeTopDown(
			function(component) {
				if(component.declaredClass == "LegendComponent"){
					if(component.mapId == mapId || component.tag == "legend"){
						legend = component;
					}
				}
			}
		);
		return legend;
	},
	
	getComponentById : function (id) {
		var comp = null;
		this.page.executeTopDown(
			function(component) {
				if(component.getId() == id){
					comp = component;
					return true;
				} else {
					return false;
				}
			}
		);
		return comp;
	},
	
	getBestMatchingComponentById : function (id) {
		var comp = null;
		this.page.executeBottomUp(
			function(component) {
				if(id.startsWith(component.getId())){
					comp = component;
					return true;
				} else {
					return false;
				}
			}
		);
		return comp;
	},

	addViewPort : function ( mapId ){
		var map = this.getMapComponent(mapId);
		map.addViewPort();
	},
	
	_createPage : function() {
		this.page = new PageComponent(this.page, "print.page");
		this._createChildren(this.page);
	},

	_createChildren : function(/*BaseComponent*/parent) {
		for(var i = 0; i < parent.children.list.length; i++){
			var child = parent.children.list[i];
			var id = parent.getId()+"."+i;
			if(child.javaClass.endsWith("MapComponent")){
				parent.children.list[i] = new MapComponent(child,id);
			} else if(child.javaClass.endsWith("PageComponent")){
				parent.children.list[i] = new PageComponent(child,id);
			} else if(child.javaClass.endsWith("LabelComponent")){
				parent.children.list[i] = new LabelComponent(child,id);
			} else if(child.javaClass.endsWith("VectorLayerComponent")){
				parent.children.list[i] = new VectorLayerComponent(child,id);
			} else if(child.javaClass.endsWith("RasterLayerComponent")){
				parent.children.list[i] = new RasterLayerComponent(child,id);
			}  else if(child.javaClass.endsWith("LegendComponent")){
				parent.children.list[i] = new LegendComponent(child,id);
			}  else if(child.javaClass.endsWith("LegendItemComponent")){
				parent.children.list[i] = new LegendItemComponent(child,id);
			}  else if(child.javaClass.endsWith("ViewPortComponent")){
				parent.children.list[i] = new ViewPortComponent(child,id);
			} else {
				parent.children.list[i] = new BaseComponent(child,id);
			}
			log.info("create component "+child.javaClass+" with id "+parent.children.list[i].id);
			this._createChildren(parent.children.list[i]);
		}
	},
	
	getPage : function() {
		return this.page;
	},
	
	getName : function() {
		return this.name;
	},
	
	setName : function(name) {
		this.name = name;
	}
		
});

File: plugin/print/print/src/main/java/org/geomajas/plugin/printing/configuration/PrintTemplate.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2015 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */
package org.geomajas.plugin.printing.configuration;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.Table;

import org.geomajas.global.Json;
import org.geomajas.plugin.printing.component.impl.PageComponentImpl;

/**
 * ???
 *
 * @author Jan De Moerloose
 */
@Entity
@Table(name = "print_template")
public class PrintTemplate {

	@Id
	@GeneratedValue(strategy = javax.persistence.GenerationType.IDENTITY)
	@Column(name = "id")
	private Long id;

	@Column(name = "name")
	private String name;

	@Column(name = "page_xml")
	private String pageXml;

	@Column(name = "template")
	private boolean template;

	private transient PageComponentImpl page;

	public static final String MAP = "map";

	public static final String LEGEND = "legend";

	public static final String PAGE = "page";

	public static final String TITLE = "title";

	public static final String DATE = "date";

	public static final String ARROW = "arrow";

	public static final String SCALEBAR = "scalebar";

	// Constructors:

	public PrintTemplate(boolean isTemplate) {
		this.template = isTemplate;
	}

	public PrintTemplate() {
		this(false);
	}

	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	@Json(serialize = false)
	public String getPageXml() {
		return pageXml;
	}

	public void setPageXml(String pageXml) {
		this.pageXml = pageXml;
	}

	public PageComponentImpl getPage() {
		return page;
	}

	public void setPage(PageComponentImpl page) {
		this.page = page;
	}

	public boolean isTemplate() {
		return template;
	}

	public void setTemplate(boolean isTemplate) {
		this.template = isTemplate;
	}

}
